### ISSUE TITLE
AutoSklearn - always producing the same output

### ISSUE THREAD

**Issue by nunocesarsa** [automl/auto-sklearn#1044]
Hello!

I have a question - one of my first times interacting through github, so let me know if questions should be somewhere else other than in "issues".

So I've been trying out AutoSklearn on my data and i always get the same outputs, variations depend only on the time budget i give to it. 

This is my model call:

`  mdl = autosklearn.regression.AutoSklearnRegressor(time_left_for_this_task=mdl_time,
                                                    per_run_time_limit=int(mdl_time/10),
                                                    seed=smac_seed,
                                                    initial_configurations_via_metalearning=38,
                                                    resampling_strategy='cv',
                                                    resampling_strategy_arguments={'folds': fold_nr},
                                                    #tmp_folder='/content/drive/MyDrive/GrassHeight/Exp02/V02/Autosklearn_dumps/'+temp_fld,
                                                    #output_folder='/content/drive/MyDrive/GrassHeight/Exp02/V02/Autosklearn_dumps/'+out_fld+'_s'+str(smac_seed)+"_t"+time_str,
                                                    #delete_output_folder_after_terminate=False,
                                                    metric = autosklearn.metrics.mean_absolute_error)`

I tested if the smac_seed varying would vary my output models and predictions but it doesn't. The model starts to overfit around 700s of training but that is, I reckon because it finds a pipeline that overfits the training data. In reality, it is at best accuracy already at about 200s. There is no variation on the predictions and error between ~200 to ~700s. 

I reckon this is because I am using the default meta-learning pipelines, correct? Meaning, once I have: initial_configurations_via_metalearning set at default, it means Autosklearn starts by evaluating from that initial set of models, hence, I end up always finding the same solution for my regression problem?

Did I understand the problem correctly here or am I missing something?

**Comment by mfeurer** [comment]
Hi there,

> I have a question - one of my first times interacting through github, so let me know if questions should be somewhere else other than in "issues".

Welcome to github then! Questions are totally fine here, we don't have the new discussions feature set up yet.

> I reckon this is because I am using the default meta-learning pipelines, correct?

I assume that this is the case. There is no randomness involved in the meta-learning, and the seed is only used for the Bayesian optimization and ensembling components of Auto-sklearn. If you run it for even longer the performance might change again if it finds an even better model. As you can observe overfitting quite early, I suggest that you increase the number of folds to obtain more reliable estimates of the pipeline quality.

**Comment by nunocesarsa** [comment]
@mfeurer thank you for the reply!

Another follow up -> does autosklearn actually do feature selection? I see it tries out different pipelines in feature pre-processing but does it try taking out feature X, or Y to see if the model improves?

Also, and this is more future research-oriented. You _et al_ have set up I think about ~40 pipelines in the meta-learning options based on kaggle data (perhaps also somewhere else). In Remote sensing, we actually use pretty much this structured data approach for classification, only more recently the CNN, RCNN is starting to gain traction. It would be cool for us in our field to set up a number of meta-models based on Sensor, bands & tasks in Remote Sensing. 

So is it possible to define these custom sets of Meta-models and plug them in into auto-sklearn currently? And if so, would be cool also to try it with autoGluon, so, do you have any idea if we can move this information around between the different automl frameworks?

**Comment by mfeurer** [comment]
> does autosklearn actually do feature selection

Yes, it includes feature selection algorithms from scikit-learn and configures them.

> does it try taking out feature X, or Y to see if the model improves?

No, it does not. Backward selection can be a very expensive procedure and is not implemented.

> So is it possible to define these custom sets of Meta-models and plug them in into auto-sklearn currently?

I assume you mean meta data for custom models? Please have a look into the scripts directory to see how meta-data is generated.

> would be cool also to try it with autoGluon, so, do you have any idea if we can move this information around between the different automl frameworks?

I don't think this is possible without diving deep into the code. Also, I'm not sure how well Auto-Gluon will work with models such as CNNs and RCNNs as the stacking and resampling procedure produces lots of models.
